### PR TITLE
chore(react-storage): refactor Refresh to DataRefresh to add composability support

### PR DESCRIPTION
**Pull request by ashwinkumar6**
#### Description of changes
- Add DataRefresh composable component
- Add DataRefreshControl and it's useDataRefresh hook
- Update LocationsView and LocationsDetailView to use the new DataRefreshControl component

#### Description of how you validated changes
Tested locally

#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added
- [x] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax
- [ ] _If this change should result in a version bump_, [changeset added](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) (This can be done after creating the PR.) This does not apply to changes made to `docs`, `e2e`, `examples`, or other private packages.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-storage/src/styles/storage-browser.css [hunk 1]
```diff
@@ -57,7 +57,7 @@
 }
 
 .storage-browser__paginate__icon,
-.storage-browser__refresh__icon,
+.storage-browser__refresh-icon,
```
Comment (ashwinkumar6) [inline packages/react-storage/src/styles/storage-browser.css:60]: Updated to follow BEM pattern. We can do this on a separate PR as well

File: packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx [hunk 2]
```diff
@@ -99,34 +91,48 @@ export function LocationsView({
...
-      <Title>Home</Title>
-      <RefreshControl
-        disableRefresh={disableRefresh}
-        handleRefresh={() => {
+  // FIXME: Eventually comes from useView hook
+  const contextValue: ControlsContext = {
+    data: {},
+    actionsConfig: {
+      type: 'SINGLE_ACTION',
+      isCancelable: true, // Question(ashwinkumar6): is refresh Cancelable ?
```
Comment (ashwinkumar6) [inline packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx:99]: QQ
is this required for refresh 
```
 data: {},
    actionsConfig: {
      type: 'SINGLE_ACTION',
      isCancelable: true, // Q
```

File: packages/react-storage/src/components/StorageBrowser/controls/__tests__/DataRefreshControl.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import { useDataRefresh } from '../hooks/useDataRefresh';
+import { DataRefreshControl } from '../DataRefreshControl';
+
+jest.mock('../hooks/useDataRefresh');
+jest.mock('../hooks/useResolvedComposable');
+
+describe('DataRefreshControl', () => {
+  // assert mocks
+  const mockUseDataRefresh = useDataRefresh as jest.Mock;
+  const mockUseResolvedComposable = useResolvedComposable as jest.Mock;
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/controls/__tests__/DataRefreshControl.spec.tsx:13]: Prefer spies over mocks. Also don't think we should mock `useResolvedComposable` since it returns the component you are testing?

File: packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx (hunk 2 above)
Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx:99]: Left some previous comments [here](https://github.com/aws-amplify/amplify-ui/pull/5939/files#diff-9cdd03a53e36958be1ed87fb20255b4814ae5e41c45be51b0062b309144da0aeR26), and [here](https://github.com/aws-amplify/amplify-ui/pull/5939/files#diff-2a298ab664ed59a235a070ed3e4ca2e5045c4383dadd8afb5ceda79cc2c142a6R159) that are somewhat related here.

The `DataRefresh` should has to do with configuration - it just renders an icon button with some static data (a11y values, etc) and accepts a handful of `props` (`isDisabled`, `children` and `onRefresh`) so there is no reason that we need to add configuration related to `actions` here.

There maybe a piece to this approach that i am jus misunderstanding but regardless it is out of scope for adding the `DataRefresh` component

**Review by calebpollman — COMMENTED** [review]
Great work @ashwinkumar6! Some feedback - please open a PR with just the component  and do the integration work/cleanup in a follow up to speed up the feedback cycle

File: packages/react-storage/src/components/StorageBrowser/controls/types.ts [hunk 4]
```diff
@@ -20,5 +20,9 @@ export interface ControlsContext {
   actionsConfig: {
     type: 'SINGLE_ACTION' | 'BATCH_ACTION';
     isCancelable: boolean;
+    dataRefresh?: {
+      onClick?: () => void;
+      disabled?: boolean;
+    };
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/react-storage/src/components/StorageBrowser/controls/types.ts:26]: Had a similar question here https://github.com/aws-amplify/amplify-ui/pull/5935/files#r1813178766. On one hand, it doesn't really matter how this gets into the controls context - but, on the other hand, I only had actionsConfig here for things mapped to action configs and would probably expect a shape more like
```ts
export interface ControlsContext {
  data: {
    taskCounts?: TaskCounts;
    isDataRefreshDisabled?: boolean;
  };
  actionsConfig: {
    type: 'SINGLE_ACTION' | 'BATCH_ACTION';
    isCancelable: boolean;
  },
  onDataRefresh?: () => void;
}
```
Does this make sense to you, @calebpollman

File: packages/react-storage/src/components/StorageBrowser/views/LocationDetailView/Controls.tsx [hunk 5]
```diff
@@ -160,21 +153,33 @@ export const LocationDetailViewControls = (): React.JSX.Element => {
...
-    <>
-      <Navigate />
-      <Title />
-      <RefreshControl
-        disableRefresh={disableRefresh}
-        handleRefresh={() => {
+  // FIXME: Eventually comes from useView hook
+  const contextValue: ControlsContext = {
+    data: {},
+    actionsConfig: {
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/react-storage/src/components/StorageBrowser/views/LocationDetailView/Controls.tsx:159]: Fwiw, not everything here really should come from the useView hook. The view components themselves will likely have work to do as well. In other words - my comment is just not entirely accurate here

File: packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx (hunk 2 above)
Comment (cshfang) [reply packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx:99]: Agree with @calebpollman here. That stuff is irrelevant on this view. It was only relevant for the StatusDisplay component because we were keying some conditional presentational logic off the action configuration. That shouldn't be applicable here

File: packages/react-storage/src/styles/storage-browser.css [hunk 6]
```diff
@@ -119,12 +119,18 @@
   margin: 0;
 }
 
-.storage-browser__refresh {
+.storage-browser__locations-view__data-refresh,
+.storage-browser__locations-detail-view__data-refresh {
```
Comment (cshfang) [inline packages/react-storage/src/styles/storage-browser.css:123]: Nit: I think we will have to re-evaluate on a whole at some point anyway so this isn't super important to get right just yet
```suggestion
.storage-browser__locations-view-data-refresh,
.storage-browser__locations-detail-view-data-refresh {
```

File: packages/react-storage/src/components/StorageBrowser/controls/types.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashwinkumar6) [reply packages/react-storage/src/components/StorageBrowser/controls/types.ts]: This type would result in
```
  const contextValue: ControlsContext = {
    data: {
      isDataRefreshDisabled: disableRefresh,
    },
    actionsConfig: {
      type: 'SINGLE_ACTION',
      isCancelable: true
    },
    onDataRefresh: () => {
      handleReset();
      handleList({
        options: { ...DEFAULT_LIST_OPTIONS, refresh: true },
      });
    },
  };
```
`actionsConfig` is still a mandatory option here. Wondering if this is relevant here

File: packages/react-storage/src/components/StorageBrowser/views/LocationDetailView/Controls.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashwinkumar6) [reply packages/react-storage/src/components/StorageBrowser/views/LocationDetailView/Controls.tsx:161]: Updated to
```
  const contextValue: ControlsContext = {
    data: {
      isDataRefreshDisabled: disableRefresh,
    },
    actionsConfig: {
      type: 'SINGLE_ACTION',
      isCancelable: true,
    },
    onDataRefresh: () => {
      if (!hasValidPath) return;
      handleReset();
      handleList({
        prefix: path,
        options: DEFAULT_REFRESH_OPTIONS,
      });
      handleLocationActionsState({ type: 'CLEAR' });
    },
  };
```

File: packages/react-storage/src/styles/storage-browser.css (hunk 6 above)
Comment (ashwinkumar6) [reply packages/react-storage/src/styles/storage-browser.css]: sounds good, updated

File: packages/react-storage/src/components/StorageBrowser/controls/__tests__/DataRefreshControl.spec.tsx (hunk 3 above)
Comment (ashwinkumar6) [reply packages/react-storage/src/components/StorageBrowser/controls/__tests__/DataRefreshControl.spec.tsx]: thx updated

File: packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx (hunk 2 above)
Comment (ashwinkumar6) [reply packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx]: discussed aync, updated to 
```
export interface ControlsContext {
  data: {
    taskCounts?: TaskCounts;
    isDataRefreshDisabled?: boolean; // <== data
  };
  actionsConfig: {
    type: 'SINGLE_ACTION' | 'BATCH_ACTION';
    isCancelable: boolean;
  },
  onDataRefresh?: () => void; // <== Handlers
}
```
With plans to refactor in upcoming PR

File: packages/react-storage/src/components/StorageBrowser/controls/DataRefreshControl.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+export const DataRefreshControl = ({
+  className,
+}: ControlProps): React.JSX.Element | null => {
+  const { props } = useDataRefresh();
+  const ResolvedDataRefresh = useResolvedComposable(DataRefresh, 'DataRefresh');
+
+  if (!props) {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (ashwinkumar6) [reply packages/react-storage/src/components/StorageBrowser/controls/DataRefreshControl.tsx]: updated

File: packages/react-storage/src/components/StorageBrowser/composables/types.ts [hunk 8]
```diff
@@ -1,7 +1,9 @@
 import { StatusDisplay } from './StatusDisplay';
+import { DataRefresh } from './DataRefresh';
 
 const composables = {
   StatusDisplay,
+  DataRefresh,
 };
```
[the remark this answers is not in the document]
Comment (ashwinkumar6) [reply packages/react-storage/src/components/StorageBrowser/composables/types.ts]: discussed async, we can perform additional cleanup in an upcoming PR

File: packages/react-storage/src/components/StorageBrowser/controls/types.ts [hunk 9]
```diff
@@ -34,13 +34,15 @@ export interface ControlsContext {
   data: {
     taskCounts?: TaskCounts;
     tableData?: TableData;
+    isDataRefreshDisabled?: boolean;
   };
-  actionsConfig: {
+  actionsConfig?: {
```
Comment (ashwinkumar6) [inline packages/react-storage/src/components/StorageBrowser/controls/types.ts:39]: Updated to optional
cc @Samaritan1011001 @calebpollman @cshfang

File: packages/react-storage/src/components/StorageBrowser/controls/hooks/useStatusDisplay.ts [hunk 10]
```diff
@@ -5,7 +5,7 @@ import { displayText } from '../../displayText/en';
 export const useStatusDisplay = (): StatusDisplayProps | null => {
   const { data, actionsConfig } = useControlsContext();
   const { taskCounts } = data;
-  const { isCancelable, type } = actionsConfig;
+  const { isCancelable, type } = actionsConfig ?? {};
```
Comment (ashwinkumar6) [inline packages/react-storage/src/components/StorageBrowser/controls/hooks/useStatusDisplay.ts:8]: required because [actionsConfig is now optional](https://github.com/aws-amplify/amplify-ui/pull/5939/files#r1817235827)

**Review by cshfang — DISMISSED** [review]

File: packages/react-storage/src/components/StorageBrowser/composables/types.ts [hunk 11]
```diff
@@ -1,9 +1,10 @@
 import { DataTableProps } from './DataTable';
 import { StatusDisplayProps } from './StatusDisplay';
-
+import { DataRefreshProps } from './DataRefresh';
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/composables/types.ts:3]:
```suggestion

import { DataRefreshProps } from './DataRefresh';
```

File: packages/react-storage/src/components/StorageBrowser/composables/types.ts [hunk 12]
```diff
@@ -1,9 +1,10 @@
...
 export interface Composables {
   DataTable: React.ComponentType<DataTableProps>;
   StatusDisplay: React.ComponentType<StatusDisplayProps>;
+  DataRefresh: React.ComponentType<DataRefreshProps>;
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/composables/types.ts:7]: Nit: alphabetize

File: packages/react-storage/src/components/StorageBrowser/controls/types.ts (hunk 9 above)
Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/controls/types.ts:39]: Still unclear why this is needed at all 😅

File: packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx [hunk 13]
```diff
@@ -102,34 +94,43 @@ export function LocationsView({
       hasError,
     });
 
+  // FIXME: Eventually comes from useView hook
+  const contextValue: ControlsContext = {
+    data: {
+      isDataRefreshDisabled: disableRefresh,
+    },
+    onDataRefresh: () => {
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx:102]: Let's update to `onRefresh`, aligning the callback names will help to prevent confusion for future maintainers

File: packages/react-storage/src/styles/storage-browser.css [hunk 14]
```diff
@@ -96,12 +96,18 @@
...
   justify-self: flex-end;
 }
 
-.storage-browser__refresh:has(~ .storage-browser__actions-menu) {
+.storage-browser__locations-view-data-refresh:has(
+    ~ .storage-browser__actions-menu
+  ),
+.storage-browser__locations-detail-view-data-refresh:has(
+    ~ .storage-browser__actions-menu
+  ) {
```
Comment (calebpollman) [inline packages/react-storage/src/styles/storage-browser.css:110]: What is happening here?

**Review by calebpollman — COMMENTED** [review]
Just a few questions/naming callouts

File: packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx [hunk 15]
```diff
@@ -102,34 +94,43 @@ export function LocationsView({
       hasError,
     });
 
+  // FIXME: Eventually comes from useView hook
```
Comment (AllanZhengYP) [inline packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx:97]: 👍. I'd lose my mind if we keep re-rendering the whole tree below.

**Review by AllanZhengYP — APPROVED** [review]

File: packages/react-storage/src/styles/storage-browser.css (hunk 14 above)
Comment (ashwinkumar6) [reply packages/react-storage/src/styles/storage-browser.css:110]: Rather than having a generic classname (`.storage-browser__refresh`) we've updated them to be view specific ([here](https://github.com/aws-amplify/amplify-ui/pull/5939/files#diff-4759f6ac7ea179a389a03d228f7fef0c5d05dacc2a7b08fcc623a0d0c4e7beb9R118) and [here](https://github.com/aws-amplify/amplify-ui/pull/5939/files#diff-2a298ab664ed59a235a070ed3e4ca2e5045c4383dadd8afb5ceda79cc2c142a6R161))
 
hence updating here to reflect the same

File: packages/react-storage/src/components/StorageBrowser/composables/types.ts (hunk 12 above)
Comment (ashwinkumar6) [reply packages/react-storage/src/components/StorageBrowser/composables/types.ts]: thx updated

File: packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx (hunk 13 above)
Comment (ashwinkumar6) [reply packages/react-storage/src/components/StorageBrowser/views/LocationsView/LocationsView.tsx]: thx updated

File: packages/react-storage/src/components/StorageBrowser/composables/types.ts (hunk 11 above)
Comment (ashwinkumar6) [reply packages/react-storage/src/components/StorageBrowser/composables/types.ts:3]: updated to alphabetize

**Review by calebpollman — APPROVED** [review]

**Review by cshfang — APPROVED** [review]
